File: kluzo-scala/src/main/scala/com/avast/kluzo/Kluzo.scala
package com.avast.kluzo

import com.avast.continuity.{Continuity, IdentityThreadNamer, ThreadNamer}

/* KEEP THIS IMPLEMENTATION IN SYNC WITH THE JAVA API VERSION */

object Kluzo {

  /** Key of the Kluzo trace ID in the Continuity context. */
  final val ContinuityKey = "traceId"

  /** HTTP header name that is used to transfer Kluzo trace ID over HTTP. */
  final val HttpHeaderName = "Avast-Kluzo-TraceId"

  /** Puts the [[com.avast.kluzo.TraceId]] into the context, names a thread and runs the given block of code
    * if traceId exists. It correctly cleans up everything after the block finishes.
    *
    * @see [[com.avast.continuity.Continuity#withContext]]
    */
  def withTraceId[A](traceId: Option[TraceId])(block: => A)(implicit threadNamer: ThreadNamer = IdentityThreadNamer): A = traceId match {
    case Some(tid) => Continuity.withContext(ContinuityKey -> tid.value)(block)
    case None => block
  }

  def getTraceId: Option[TraceId] = Continuity.getFromContext(ContinuityKey).map(TraceId.apply)

  def setTraceId(traceId: TraceId): Unit = Continuity.putToContext(ContinuityKey, traceId.value)

}

File: kluzo/src/main/java/com/avast/kluzo/javaapi/Kluzo.java
package com.avast.kluzo.javaapi;

import com.avast.continuity.ThreadNamer;
import com.avast.continuity.javaapi.Continuity;
import com.avast.kluzo.Kluzo$;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.Callable;

/* KEEP THIS IMPLEMENTATION IN SYNC WITH THE SCALA API VERSION */

public final class Kluzo {

    /**
     * Key of the Kluzo trace ID in the Continuity context.
     */
    public static final String CONTINUITY_KEY = Kluzo$.MODULE$.ContinuityKey();

    /**
     * HTTP header name that is used to transfer Kluzo trace ID over HTTP.
     */
    public static final String HTTP_HEADER_NAME = Kluzo$.MODULE$.HttpHeaderName();

    private Kluzo() {
    }

    /**
     * Puts the {@link TraceId} into the context, names a thread and runs the given block of code
     * if traceId exists. It correctly cleans up everything after the block finishes.
     *
     * @see com.avast.continuity.javaapi.Continuity#withContext(Callable)
     */
    public static <T> T withTraceId(Optional<TraceId> traceId, Callable<T> block) throws RuntimeException {
        if (traceId.isPresent()) {
            Map<String, String> ctxValues = new HashMap<>();
            ctxValues.put(CONTINUITY_KEY, traceId.get().getValue());
            return Continuity.withContext(ctxValues, block);
        } else {
            try {
                return block.call();
            } catch (Exception ex) {
                throw new RuntimeException(ex);
            }
        }
    }

    /**
     * Puts the {@link TraceId} into the context, names a thread and runs the given block of code
     * if traceId exists. It correctly cleans up everything after the block finishes.
     *
     * @see com.avast.continuity.javaapi.Continuity#withContext(Callable)
     */
    public static <T> T withTraceId(Optional<TraceId> traceId, ThreadNamer threadNamer, final Callable<T> block) throws RuntimeException {
        if (traceId.isPresent()) {
            Map<String, String> ctxValues = new HashMap<>();
            ctxValues.put(CONTINUITY_KEY, traceId.get().getValue());
            return Continuity.withContext(ctxValues, threadNamer, block);
        } else {
            try {
                return block.call();
            } catch (Exception ex) {
                throw new RuntimeException(ex);
            }
        }
    }

    public static Optional<TraceId> getTraceId() {
        return Continuity.getFromContext(CONTINUITY_KEY).map(TraceId::new);
    }

    public static void setTraceId(TraceId traceId) {
        Continuity.putToContext(CONTINUITY_KEY, traceId.getValue());
    }

}
